[#3] 회원가입 기능

choitaehoon opened the pull request:
- 회원가입 구현, 테스트 코드 작성
- 추후에 Global Exception Handling 검토

choitaehoon committed:
[#3] 회원가입 기능

- 회원가입 구현, 테스트 코드 작성
- 추후에 Global Exception Handling 검토

f-lab-dev commented on src/main/java/com/social/instagram/controller/AccountController.java:
```diff
@@ -0,0 +1,38 @@
+
+    private final AccountService accountService;
+
+    @PostMapping("register")
```
HTTP에서 `POST`는 어떤 메소드인가요? `register`라는 말과 겹치지 않나요?

choitaehoon replied:
HTTP POST메소드는 데이터 전송을 기반으로 한 요청 메소드 입니다. POST 방식은 GET방식과는 다르게 URL에 붙여서 보내지 않고 BODY에다가 데이터를 넣어서 보냅니다. 지금 생각해보니 POST와 register의 의미가 겹치는것 같습니다. 수정하겠습니다.

f-lab-dev commented on src/main/java/com/social/instagram/controller/AccountController.java:
```diff
@@ -0,0 +1,38 @@
+    public ResponseEntity<Void> accountRegister(@Valid @RequestBody AccountDto account) {
+        accountService.accountRegister(Account.changeAccountEntity(account));
+
+        return new ResponseEntity<>(HttpStatus.CREATED);
```
이 메소드는 항상 같은 값을 리턴합니다. 그러면 매번 같은 객체를 계속 생성해줘야할 이유가 있을까요?
상수로 리턴해줘도 되지 않을까요?

choitaehoon replied:
매번 같은 객체를 생성해주는 것보다 상수로 리턴하여 재사용 할 수 있게 하는 것이 효율적이라고 생각이 드네요.

f-lab-dev commented on src/main/java/com/social/instagram/domain/Account.java:
```diff
@@ -0,0 +1,65 @@
+    기본 생성자를 자동으로 생성
+*/
+@Entity
+@NoArgsConstructor
```
기본생성자를 public으로 생성해주면 누구나 빈 객체를 생성할 수 있을 것 같은데요.
완전 비어있는 객체를 만들어줘야할 이유가 있을까요?

choitaehoon replied:
JPA를 이용하여 엔티티를 작성하는 경우 비어있는 생성자를 만들어야 할 때가 있습니다.
비어있는 생성자를 만들어야 할 때는 파라미터가 있는 생성자를 정의 할 때 입니다.
왜냐하면, 리플렉션을 이용해 해당 객체를 생성 할 때 기본생성자를 호출하게되는데 파라미터가 있는 생성자를 정의 한다면 해당 엔티티의 객체를 만들 수 없기 때문입니다.

choitaehoon replied:
기본생성자를 프로젝트 코드상으로 선언 할 수 있으니 `@NoArgsConstructor(AccessLevel.PROTECTED)`를 이용하는 것이 더 좋다고 판단이 들어 적용하겠습니다.

f-lab-dev commented on src/main/java/com/social/instagram/exception/DuplicateUserIdException.java:
```diff
@@ -0,0 +1,9 @@
+package com.social.instagram.exception;
+
+public class DuplicateUserIdException extends RuntimeException {
```
유저가 잘못된 아이디를 입력한 것이므로 `IllegalArgumentException`을 상속받게 해주면 어떨까요?

f-lab-dev commented on src/main/java/com/social/instagram/service/AccountService.java:
```diff
@@ -0,0 +1,27 @@
+
+    public void checkDuplicateUserId(String userId) {
+        if (accountRepository.existsByUserId(userId)) {
+            throw new DuplicateUserIdException("계정이 존재합니다");
```
- 예외를 던지는것도 비용이 비쌉니다. 굳이 예외를 던지는것보다 controller에서 계정 유무를 확인한 후에 status code로 리턴해주는게 더 낫지 않을까요?
- 이 메소드는 계정이 없으면 예외를 던지는것보다 계정 유무 여부를 반환해주는 메소드로 만들면 이곳저곳에서 재사용성이 더 좋아질 것 같네요~

choitaehoon replied:
예외를 생성, 경로를 추적하기 때문에 비용이 비싸군요. 서비스를 확장하게 되면 지금 보다 복잡도가 높아져 stack depth가 더 깊어지기 때문에 계정 유무 확인 후 중복된 아이디가 있다면 status code로 리턴해주는것이 합리적이라고 판단이 들어 적용하겠습니다.

choitaehoon replied:
생각해보니 controller에서 계정 유무를 파악하는 것 보다 aop를 이용하여 회원등록, 예외처리를 분리하는게 괜찮지 않을까 생각합니다. 
1. controller안의 if/else if depth 깊어질 가능성(현재는 중복된 아이디만 체크하고 있지만 서비스가 확장함에 따라 로직이 증가하는 문제점)
2. 회원 등록과 예외 처리 성격을 분리하려는 의도

f-lab-dev commented on src/main/java/com/social/instagram/util/PasswordEncodingUtil.java:
```diff
@@ -0,0 +1,55 @@
+import java.security.NoSuchAlgorithmException;
+import java.util.Random;
+
+public class PasswordEncodingUtil {
```
- 이 클래스는 암호화 알고리즘 선택쪽에서 확장의 가능성이 있어보입니다. 이렇게 확장이 불가능한 구조로 만든 이유가 있으신가요?
- 암호화 방식이 아래에 하드코딩 되어있습니다. 상수로 빼면 어떨까요?
- 이 암호화는 여러가지로 재사용가능할 것 같은데 이름때문에 password쪽으로만 역할이 한정될 것 같습니다

choitaehoon replied:
암호화 알고리즘 확장 가능성을 열어두는것이 좋다고 생각이 드네요. 전반적으로 개선하겠습니다.

f-lab-dev commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,25 @@
+    driver-class-name: com.mysql.cj.jdbc.Driver
+    url: jdbc:mysql://localhost:3306/instagram?serverTimezone=UTC&characterEncoding=UTF-8
+    username: root
+    password: chlxogns12
```
이것은 자주 사용하시는 패스워드같은데 이런 공개된 저장소에 올리셔도 되나요?

choitaehoon replied:
자주 사용하는 패스워드가 아니기 때문에 올렸습니다. 하지만 보안 측면에서 운영 환경을 숨기도록 수정하겠습니다.

f-lab-dev replied:
만약 올려도 된다고 하면 로컬용 property를 분리하기 위해 `application-default.properties`에 써주시면 좋겠네요

f-lab-dev commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,25 @@
+    properties:
+      hibernate:
+        #hibernate가 DB에 날리는 모든 쿼리 출력
+        show_sql: true
```
이 옵션은 production까지도 적용될 것 같은데요. 불필요한 로그가 가득 생성되지 않을까요?
실서버와 설정을 분리하려면 어떻게 해야할까요?

choitaehoon replied:
production 옵션과 local 옵션을 분리해야 될거같네요. 실 서버를 운영하다가 불필요한 로그로 인해 용량이 꽉차 서버 다운될 가능성이 있을거 같습니다.
실서버와 설정을 분리하기 위해 spring.profiles.active를 적용하겠습니다

choitaehoon committed:
회원가입 리뷰 반영

- AccountController의 accountRegister api 이름 변경
- ResponseEntity 리턴 값 상수 적용
- NoArgsConstructor(AccessLevel.PROTECTED) 적용
- application.yml 로컬/운영 환경설정 분리
- 암호화 확장

f-lab-dev commented on src/main/java/com/social/instagram/controller/AccountController.java:
```diff
@@ -28,11 +29,12 @@
+        accountService.accountRegister(Account.changeAccountEntity(account, encryptionFactory.sha256Util()));
 
-        return new ResponseEntity<>(HttpStatus.CREATED);
+        return RESPONSE_ENTITY;
```
이 상수의 이름에서 역할을 유추할 수 있을까요? `CREATE` 등의 이름으로 역할을 명확히 해주는게 좋지 않을까요?

choitaehoon replied:
`CREATE` keword를 빼먹었네요. `CREATE`이름으로 역할을 명확하게 해주는 것이 좋을거 같습니다

f-lab-dev commented on src/main/java/com/social/instagram/controller/AccountController.java:
```diff
@@ -28,11 +29,12 @@
+    @PostMapping("sign-up")
     public ResponseEntity<Void> accountRegister(@Valid @RequestBody AccountDto account) {
-        accountService.accountRegister(Account.changeAccountEntity(account));
+        EncryptionFactory encryptionFactory = new EncryptionFactory();
```
메소드가 호출될때마다 `EncryptionFactory`가 계속 생성될 것 같은데요. 이 클래스의 객체는 재사용이 가능해보입니다. bean으로 등록해서 돌려쓰는게 어떨까요?

choitaehoon replied:
구현 할 당시 `Sha256Util` 클래스가 state하여 thread-unsafe를 예상했기 때문에 bean으로 등록을 안했습니다. 하지만 다시 생각해보니 stateless 할거 같네요. bean으로 등록하는게 맞다고 생각합니다

f-lab-dev commented on src/main/java/com/social/instagram/util/encoding/Sha256Util.java:
```diff
@@ -0,0 +1,61 @@
+import java.security.NoSuchAlgorithmException;
+import java.util.Random;
+
+public class Sha256Util implements EncryptionUtil {
```
이름은 Util에서 Service로 바꿔줘도 되지 않을까요?

choitaehoon replied:
service는 여러 DAO에 대한 데이터 접근/갱신한 데이터를 기반으로 비즈니스 로직을 수행이기 때문에 service보단 Util의 성격이 근접하지 않나요?

choitaehoon replied:
Sha256 알고리즘을 제공하는 성격 가지고 있어 Util에 적합하지 않나요?

choitaehoon replied:
sha256 알고리즘을 제공한다는 의미에서 Util의 naming이 낫지 않을까 생각합니다

f-lab-dev commented on src/main/java/com/social/instagram/util/encoding/Sha256Util.java:
```diff
@@ -0,0 +1,61 @@
+        return saltBuilder.toString();
+    }
+
+    private static String getEncrypt(String password, String salt) {
```
이 메소드는 sha256 알고리즘에 종속적인가요?

choitaehoon replied:
salt와 sha256 알고리즘을 독립적으로 사용하지 않고  salt + sha256을 조합한 메서드를 사용할거라 예상해서 종속적으로 구현했습니다.

f-lab-dev commented on src/main/java/com/social/instagram/factory/EncryptionFactory.java:
```diff
@@ -0,0 +1,11 @@
+public class EncryptionFactory {
+
+    public Sha256Util sha256Util() {
+        return new Sha256Util();
```
이 `Sha256Util`도 stateless해서 객체를 돌려쓸 수 있을텐데 매번 새 객체를 생성해주는게 맞는걸까요?

choitaehoon replied:
`Sha256Util`클래스가 state한걸로 잘못생각한거 같네요. Bean으로 등록해도 thread-safe 하겠네요

choitaehoon committed:
회원가입 리뷰 반영

- ResponseEntity 리턴 값 명확한 이름으로 변경
- ResponseEntity error, success 분리(유지보수성)
- Exception Handling 적용
- EncryptionFactory bean 등록
- 추후에 sha256Util, sha256Service naming 결정

f-lab-dev commented on src/main/java/com/social/instagram/exception/DuplicateUserIdException.java:
```diff
@@ -1,9 +1,5 @@
-    public DuplicateUserIdException(String message) {
-        super(message);
-    }
+public class DuplicateUserIdException extends IllegalArgumentException{
```
여기도 띄어쓰기를 하나 해주는게 좋겠네요. `intellij checkstyle`을 검색해보시고, 적절한 컨벤션을 만들어서 써보셔도 좋을 것 같아요

```suggestion
public class DuplicateUserIdException extends IllegalArgumentException {
```

f-lab-dev commented on src/main/java/com/social/instagram/util/httpstatus/ResponseSuccessContents.java:
```diff
@@ -3,9 +3,9 @@
 import org.springframework.http.ResponseEntity;
 
-public class ResponseContents {
+public class ResponseSuccessContents {
```
굳이 성공에 관련된것만 모아두시는 이유가 있을까요? 그리고 상수 모음이니 `Contents`가 아니라 `Constants`로 지어주면 좋을 것 같아요

choitaehoon replied:
body가 있는 것과 없는 것을 구분 하기 위해 나눴는데 naming을 잘못지었네요. 수정하겠습니다

choitaehoon committed:
회원가입 리뷰 반영

- CheckStyle 적용
- ResponseConstants naming 변경

f-lab-dev commented on src/main/java/com/social/instagram/util/httpstatus/ResponseBodyNoExistenceConstants.java:
```diff
@@ -3,7 +3,7 @@
 import org.springframework.http.ResponseEntity;
 
-public class ResponseSuccessContents {
+public class ResponseBodyNoExistenceConstants {
```
그냥 `ResponseConstants` 정도로 지어주면 될 것 같은데 이렇게 세세한 이름으로 지어주신 이유가 있으실까요?

choitaehoon replied:
Response에 대한 상수가 많아지게 되면 body있는 것과 body가 없는게 섞여 가독성이 떨어질거 같아 세세하게 이름을 지었습니다.

f-lab-dev replied:
body가 있는 것은 상수로 만들 필요가 있을까요? body가 있는 response들은 그때마다 계속 값이 달라질 것 같은데요

choitaehoon replied:
동적으로 body가 달라질 경우가 있을 수 있겠네요. `ResponseConstants ` naming이 괜찮겠군요

choitaehoon committed:
회원가입 리뷰 반영

- ResponseConstants로 naming 변경

choitaehoon committed:
회원가입 기능

- AccountDto message 수정
- MethodArgumentNotValidException handler 추가
- ErrorMessage 생성

choitaehoon committed:
회원가입 기능

- ErrorMessage 가독성 증가 및 부연 설명

choitaehoon committed:
회원가입 기능

- ErrorMessage 줄바꿈 문자 상수 적용

f-lab-dev commented on src/main/java/com/social/instagram/validation/ErrorMessage.java:
```diff
@@ -22,6 +22,7 @@
 public class ErrorMessage {
 
     private final List<FieldError> errorMessage;
+    private static final String LINE_BREAK = "\n";
```
이런것은 다른곳에 어디 미리 정해져있는 상수가 없을까요?

choitaehoon replied:
`LINE_BREAK`의 상수를 따로 다른곳에 정하지 않은 이유는 아래와 같이 생각했습니다
* 에러메시지를 조작하는 과정에서 상수가 많이 늘어날거 같지 않다는 생각
* 상수가 많이 늘어나지 않는 상황에서 상수에 관한 파일을 따로 생성하는 것도 관리 비용을 증가시킬 수 있는 점

choitaehoon committed:
회원가입 기능

- EncryptionEncoding, Sha256Encoding으로 naming 변경
- Encoding에 관한 상수 분리
- ErrorMessage 주석 수정

f-lab-dev commented on src/main/java/com/social/instagram/domain/Account.java:
```diff
@@ -1,7 +1,7 @@
 
 import com.social.instagram.dto.AccountDto;
-import com.social.instagram.util.encoding.EncryptionUtil;
+import com.social.instagram.util.encoding.EncryptionEncoding;
```
`EncryptionEncoding`에서 `Encoding`은 현재 진행형입니다. 행위를 표현하는 단어이기 때문에 클래스 이름으로 적절해보이지 않네요~

f-lab-dev commented on src/main/java/com/social/instagram/util/encoding/EncodingConstants.java:
```diff
@@ -0,0 +1,17 @@
+package com.social.instagram.util.encoding;
+
+public class EncodingConstants {
```
`Encoding`은 엄청 여러가지가 있습니다. 명확하게 암호화라는 뜻으로 `Encrypt`가 나을 것 같네요

choitaehoon replied:
`security`, `strategy` 단어도 한번 생각해봤는데 가장 좋을만한 naming이 저도 `Encryption` 인거 같네요

choitaehoon committed:
회원가입 리뷰 반영

- 암호화 naming Encryption으로 변경
